# the code of analyses of NeuN project
## FIRE folder contains the code of FIREs analysis
1. identification of FIREs and super-FREs
2. venn diagram for NeuN+ and NeuN- FIREs
3. differential FIREs definition and volcano plot for differential FIREs between NeuN+ and NeuN- FIREs cells
4. barplots for differential FIREs overlapping with differential H3K27ac ChIP-seq between NeuN+ and NeuN- cells
5. two representative plots of differential FIREs overlapping with gene promoters and boxplots for gene expression between NeuN+ and NeuN-    cells.
6. barplots for GO analyses of genes aggsigned to differential FIREs in NeuN+ and NeuN- cells
7. boxplot for expression levels of genes assigned to differential FIREs in NeuN+ and NeuN- cells
8. EWCE analysis for genes assigned to differential FIREs in NeuN+ and NeuN- cells

## superFIRE folder contains the code of super-FIREs analysis
1. identification of FIREs and super-FREs
2. venn diagram for NeuN+ and NeuN- super-FIREs
3. barplots for super-FIREs overlapping with differential H3K27ac ChIP-seq and gene promoters in NeuN+ and NeuN- cells
4. two representative plots of super-FIREs overlapping with gene promoters and boxplots for gene expression between NeuN+ and NeuN-          cells.
5. barplots for GO analyses of genes aggsigned to super-FIREs in NeuN+ and NeuN- cells
6. boxplot for expression levels of genes assigned to super-FIREs in NeuN+ and NeuN- cells

## NeuN_Loops folder contains the code of enhancer-promoter interactions analyses
1. barplot for loops within TADs and promoter-promoter interactions
2. barplot for the number of Loops with the distance of loops
3. barplot for the number of genes with the number of enhancers
4. boxplots for gene expression with the number of enhancers interacting with promoters
5. barplot for loops overlapping with hg19 chromHMM
6. motif analysis and expression levels of corresponding TFs
7. two representative plots of enhancer-promoter interactions and boxplots for gene expression between NeuN+ and NeuN- cells.
8. enrichment analyses of genes assigned to enhancer-promoter interactions in developmental and synaptic module
9. heatmap for expression levels of genes assigned to enhancer-promoter interactions in NeuN+ and NeuN- cells

## AD_acetylation folder contains the code of genes subjected to dysacetylation in AD
1. barplot for genes assigned to enhancer-promoter interactions subjected to dysregulation in AD
2. venn diagram for genes identified in step1 between NeuN+ and NeuN- cells
3. two representative plots of genes assigned to enhancer-promoter interactions and subjected to dysregulation in AD and boxplots for gene expression between NeuN+ and NeuN- cells.
4. heatmap for expression levels of genes assigned to enhancer-promoter interactions and subjected to dysregulation in AD
5. module enrichment analysis for genes assigned to enhancer-promoter interactions and subjected to dysregulation in AD

## AD_GWAS folder contains the code of AD risk genes
1. LDSC analysis for differential H3K27ac CHIP-seq and barplot for heritability enrichment of AD risk variants between Neu+ and NeuN- cells.
2. one repesentative plot for one variant within the enhancer-promoter interaction
3. barplots for GO analyses of AD risk genes idnetified in NeuN- cells
4. boxplot for developmental expresison of AD risk genes idnetified in NeuN- cells
5. heatmap of cellular expresison of AD risk genes idnetified in NeuN- cells
6. enrichment analysis for AD risk genes idnetified in NeuN- cells

## Glu_GABA folder contains the code for Glu and GABA loops analysis
1. venn diagram for genes assigned to enhancer-promoter interactions shared between Glu and GABA cells
2. heatmap for expression levels of genes assigned to enhancer-promoter interactions shared between Glu and GABA cells
3. two representative plots of genes assigned to enhancer-promoter interactions and boxplots for gene expression between Glu and GABA cells.

## Glu_GABA folder contains the code for Glu and GABA loops analysis
1. LDSC analysis for differential H3K27ac CHIP-seq and barplot for heritability enrichment of SCZ and BD risk variants between Glu and GABA cells.
2. heatmap for expression levels of SCZ and BD riskgenes assigned to enhancer-promoter interactions shared between Glu and GABA cells
